stdlib/public/core/CollectionDifference.swift

```diff
@@ -0,0 +1,353 @@
+      return lhs.i < rhs.i
+    }
+    
+    let i: Int
+    init(_ index: Int) {
+      i = index
+    }
+  }
+
+  public var startIndex: CollectionDifference<ChangeElement>.Index {
+    return Index(0)
+  }
+  
+  public var endIndex: CollectionDifference<ChangeElement>.Index {
+    return Index(removals.count + insertions.count)
+  }
+  
+  public func index(after index: CollectionDifference<ChangeElement>.Index) -> CollectionDifference<ChangeElement>.Index {
+    return Index(index.i + 1)
+  }
+  
+  public subscript(position: CollectionDifference<ChangeElement>.Index) -> Element {
+    return position.i < removals.count ? removals[removals.count - (position.i + 1)] : insertions[position.i - removals.count]
+  }
+  
+  public func index(before index: CollectionDifference<ChangeElement>.Index) -> CollectionDifference<ChangeElement>.Index {
+    return Index(index.i - 1)
+  }
+  
+  public func formIndex(_ index: inout CollectionDifference<ChangeElement>.Index, offsetBy distance: Int) {
+    index = Index(index.i + distance)
+  }
+  
+  public func distance(from start: CollectionDifference<ChangeElement>.Index, to end: CollectionDifference<ChangeElement>.Index) -> Int {
+    return end.i - start.i
+  }
+}
+
+extension CollectionDifference.Change: Equatable where ChangeElement: Equatable {}
+
+extension CollectionDifference: Equatable where ChangeElement: Equatable {}
+
+extension CollectionDifference.Change: Hashable where ChangeElement: Hashable {}
+
+extension CollectionDifference: Hashable where ChangeElement: Hashable {
+
+  /// Infers which `ChangeElement`s have been both inserted and removed only
+  /// once and returns a new difference with those associations.
+  ///
+  /// - Returns: an instance with all possible moves inferred.
+  ///
+  /// - Complexity: O(*n*) where *n* is `self.count`
+  public func inferringMoves() -> CollectionDifference<ChangeElement> {
+    let removeDict: [ChangeElement:Int?] = {
+      var res = [ChangeElement:Int?](minimumCapacity: Swift.min(removals.count, insertions.count))
+      for r in removals {
+        let element = r.element
+        if res[element] != .none {
+          res[element] = .some(.none)
+        } else {
+          res[element] = .some(r.offset)
+        }
+      }
+      return res.filter { (_, v) -> Bool in v != .none }
+    }()
+    
+    let insertDict: [ChangeElement:Int?] = {
+      var res = [ChangeElement:Int?](minimumCapacity: Swift.min(removals.count, insertions.count))
+      for i in insertions {
+        let element = i.element
+        if res[element] != .none {
+          res[element] = .some(.none)
+        } else {
+          res[element] = .some(i.offset)
+        }
+      }
+      return res.filter { (_, v) -> Bool in v != .none }
+    }()
+
+    return CollectionDifference.init(validatedChanges:map({ (c: CollectionDifference<ChangeElement>.Change) -> CollectionDifference<ChangeElement>.Change in
```

Comment:
The `Change` type can be used unqualified within `CollectionDifference`, which should help making this a lot more readable:

```suggestion
    return CollectionDifference(validatedChanges: map { (change: Change) -> Change in
```

I'm pretty sure we don't need provide the closure's argument/return types explicitly, either.